In [4]:
import geopandas
import geopy
import pandas as pd

In [281]:
from geopy.geocoders import Nominatim

# create a Nominatim object
geolocator = Nominatim(user_agent="y'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'")
location = geolocator.geocode("121 N LaSalle St, Chicago")

In [ ]:
import pandas as pd

# these are the top 5 rated high schools in Chicago
top5 = ['г.Есик, Фрунзе, 16'
       ]

df = pd.DataFrame(top5, columns=['school'])
df['location'] = df['school'].apply(lambda x: geolocator.geocode(x))

In [360]:
bin_sample = pd.read_csv('bin_add_region.csv', index_col=0, dtype={
                    'bin': 'str'})
bin_sample.sample(3)

bin                                         full_add  \
96684  810121402951                              АКТОБЕ,Уалиханова,1   
80561  750110403252  г. Нур-Султан,р-н Алматы,Рақымжан Қошқарбаев,15   
91226  781129300681              УСТЬ-КАМЕНОГОРСК,Мелиораторов,14(2)   

       KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode  
96684                                                4.0       
80561                                                1.0       
91226                                               16.0

In [367]:
bin_sample.sample(10)

bin                                           full_add  \
106920  850217300195       обл. Атырауская,г. Атырау,ул. Сатпаева,д. 30   
125039  940210451325  обл. Мангистауская,с. Жетибай,Казак Аулы,дом 19 б   
60478   680101414680  обл. Жамбылская,г. Тараз,мас.Ветеран ул.Долана,54   
31670   560809301686  обл. Западно-Казахстанская,г. Уральск,Т.Масина,53   
24074   211040006867  г. Нур-Султан,р-н Алматы,ул. Габидена Мустафин...   
11178   130640001540               р-он Байконыр,Амангелды Иманов,уч.22   
122273  920223350853  обл. Восточно-Казахстанская,г. Усть-Каменогорс...   
122407  920325300198                 обл. Жамбылская,с. Аухатты,Абая,69   
57379   661119300036                          АКТОБЕ,Санкибай-батыра,26   
64443   690520401866    обл. Кызылординская,пос. Айтеке би,Айтеке би,бн   

        KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode  
106920                                                6.0       
125039                                                NaN       
60478                                                 8.0       
31670                                                 7.0       
24074                                                 1.0       
11178                                                 1.0       
122273                                               16.0       
122407                                                8.0       
57379                                                 NaN       
64443                                                11.0

In [366]:
bin_sample[bin_sample['bin'] == '740329400346']['full_add']

78384    обл. Костанайская,с. Силантьевка,ул. Омара Шип...
Name: full_add, dtype: object

In [302]:
bin_sample = bin_sample[bin_sample['KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode']==1]

In [303]:
bin_sample

bin                                    full_add  \
9       000111650722                        АСТАНА,Республика,37   
27      000127601781     г. Нур-Султан,р-н Есиль,ул.Бокейхана,15   
32      000140000240       г. Нур-Султан,р-н Байконыр,Иманова,12   
97      000220551251    г. Нур-Султан,р-н Байконыр,Алаш шоссе,32   
118     000240001926             р-он Байконыр,ул. Жиенкулова,16   
...              ...                                         ...   
131069  991209451470    г. Нур-Султан,р-н Алматы,Кошкарбаева,1/1   
131082  991224400041  г. Нур-Султан,р-он Есиль,проспект Достык,9   
131096  991240000717    г. Нур-Султан,р-н Сарыарка,Есенберлина,8   
131112  991240002723                САРЫАРКА,ул.Суйінбай ақын,81   
131154  991241014930                   р-он Байконыр,Кенесары,45   

        KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode  
9                                                     1.0       
27                                                    1.0       
32                                                    1.0       
97                                                    1.0       
118                                                   1.0       
...                                                   ...       
131069                                                1.0       
131082                                                1.0       
131096                                                1.0       
131112                                                1.0       
131154                                                1.0       

[8313 rows x 3 columns]

In [304]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace('р-н', "район")
bin_sample["full_add"] = bin_sample["full_add"].str.replace('р-он', "район")
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\г+\.', "")
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\ул+\.', "")
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\пр+\.', "")
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\д+\.', "")
bin_sample["full_add"] = bin_sample["full_add"].str.replace('мкр.', "микрорайон ", case=False)
bin_sample['full_add'] = bin_sample['full_add'].apply(lambda x: re.sub(r'(\d)\D+$', r'\1', x))

In [305]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\г+\.', "")

<ipython-input-305-5cbc6289fa22>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\г+\.', "")


In [306]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\ул+\.', "")

<ipython-input-306-e60916c2f37c>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\ул+\.', "")


In [307]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\пр+\.', "")

<ipython-input-307-618707c6b653>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\пр+\.', "")


In [308]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\д+\.', "")

<ipython-input-308-cbbfc90597e6>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample["full_add"] = bin_sample["full_add"].str.replace(r'\д+\.', "")


In [309]:
bin_sample["full_add"] = bin_sample["full_add"].str.replace('мкр.', "микрорайон ", case=False)

<ipython-input-309-b3dc3a4a52ec>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample["full_add"] = bin_sample["full_add"].str.replace('мкр.', "микрорайон ", case=False)


In [310]:
bin_sample['full_add'] = bin_sample['full_add'].apply(lambda x: re.sub(r'(\d)\D+$', r'\1', x))

In [311]:
bin_sample

bin                                   full_add  \
9       000111650722                       АСТАНА,Республика,37   
27      000127601781        Нур-Султан,район Есиль,Бокейхана,15   
32      000140000240       Нур-Султан,район Байконыр,Иманова,12   
97      000220551251    Нур-Султан,район Байконыр,Алаш шоссе,32   
118     000240001926              район Байконыр, Жиенкулова,16   
...              ...                                        ...   
131069  991209451470    Нур-Султан,район Алматы,Кошкарбаева,1/1   
131082  991224400041   Нур-Султан,район Есиль,проспект Достык,9   
131096  991240000717    Нур-Султан,район Сарыарка,Есенберлина,8   
131112  991240002723                  САРЫАРКА,Суйінбай ақын,81   
131154  991241014930                 район Байконыр,Кенесары,45   

        KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode  
9                                                     1.0       
27                                                    1.0       
32                                                    1.0       
97                                                    1.0       
118                                                   1.0       
...                                                   ...       
131069                                                1.0       
131082                                                1.0       
131096                                                1.0       
131112                                                1.0       
131154                                                1.0       

[8313 rows x 3 columns]

In [312]:
from geopy.extra.rate_limiter import RateLimiter

# 1 - conveneint function to delay between geocoding calls
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)

In [344]:
bin_sample.shape

(8313, 3)

In [349]:
bin_sample_9 = bin_sample[8001:8313]

In [350]:
import time

start = time.time()
bin_sample_9['location'] = bin_sample_9['full_add'].apply(lambda x: geocode(x))

end = time.time()
print(end - start)

311.9047601222992


<ipython-input-350-cfaa425ba8c0>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_9['location'] = bin_sample_9['full_add'].apply(lambda x: geocode(x))


In [351]:
bin_sample_9[bin_sample_9['location'].notna()].shape

(189, 4)

In [354]:
bin_location = pd.concat([bin_sample_1, bin_sample_2, bin_sample_3, bin_sample_4, bin_sample_5, bin_sample_6,
          bin_sample_7, bin_sample_8,bin_sample_9])

In [358]:
bin_location[bin_location['location'].isna()].sample(20)

bin                                           full_add  \
115410  880725451341                             район Байконыр,Алаш,15   
8765    110240006662    Нур-Султан,район Алматы,проспект Тауелсиздик,34   
81744   750615402270                                АСТАНА,Иманбаева,,8   
90493   780821401247            Нур-Султан,район Байконыр,Шоссе алаш,15   
30625   550925400653             Нур-Султан,район Сарыарка,Ақан Сері,16   
75132   730125402867             с. Андас Батыра, Свеклопунктовская, 41   
59134   670702350060        Нур-Султан,район Байконыр, Шоссе Алаш, 11/1   
131000  991041001078        Нур-Султан,район Есиль,Қайым Мұхамедханов,9   
101530  821231450032              Нур-Султан,район Байконыр,Кенесары,69   
126385  950326450459   Нур-Султан,район Сарыарка,Улица Бухарбай баты...   
130164  990140005355                        Жамбылский район,Иванова,24   
41830   610813301352                          АСТАНА,Ш.Кудайбердыулы,38   
124977  940128351383          Нур-Султан,район Есиль,Шоссе Коргальжын,2   
22022   200640011484       Нур-Султан,район Есиль,коргалжинское шоссе,5   
113110  870820350827               Нур-Султан,район Байконыр,Акжол,27/2   
19990   190440007751                      район Байконыр,ул Иманбаева,8   
98818   811125450103         Нур-Султан,район Байконыр, им. Ташенова,17   
71133   710912401338            Нур-Султан,район Есиль,Туркестан 4,нп-1   
61145   680322300219              Нур-Султан,район Сарыарка, Победы, 77   
87797   770825400328          Нур-Султан,район Сарыарка, Б. Богенбая,63   

        KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode location  
115410                                                1.0          None  
8765                                                  1.0          None  
81744                                                 1.0          None  
90493                                                 1.0          None  
30625                                                 1.0          None  
75132                                                 1.0          None  
59134                                                 1.0          None  
131000                                                1.0          None  
101530                                                1.0          None  
126385                                                1.0          None  
130164                                                1.0          None  
41830                                                 1.0          None  
124977                                                1.0          None  
22022                                                 1.0          None  
113110                                                1.0          None  
19990                                                 1.0          None  
98818                                                 1.0          None  
71133                                                 1.0          None  
61145                                                 1.0          None  
87797                                                 1.0          None

In [327]:
bin_sample_2.sample(10)

bin                                         full_add  \
13659  150640013726             Нур-Султан,район Байконыр,Пушкина,46   
18857  180940017269                        АСТАНА,Кабанбай батыра,62   
12454  140640021001        Нур-Султан,район Байконыр,Сейфуллина,47/2   
19334  181240011397                        АСТАНА,Проспект Достык,18   
16286  170340027253          АСТАНА,Индустриальный парк,  101,стр.30   
19463  190140008675                         район Алматы, Сатпаева,5   
12992  141240005167                             АСТАНА,Республики,21   
13531  150540001937   Нур-Султан, Нур-султан,Жанайдара Жирентаева,10   
11837  131240023242                      район Алматы, Куйши Дина,28   
13615  150542017141            АСТАНА, Джангильдина,  19, каб. 31-33   

       KkmsRetailPlaceOrgCatalog_retailPlaceAddressRegionCode  \
13659                                                1.0        
18857                                                1.0        
12454                                                1.0        
19334                                                1.0        
16286                                                1.0        
19463                                                1.0        
12992                                                1.0        
13531                                                1.0        
11837                                                1.0        
13615                                                1.0        

                                                location  
13659                                               None  
18857  (Мега Силк Вей, 62, проспект Кабанбай Батыра, ...  
12454                                               None  
19334                                               None  
16286                                               None  
19463  (5, Сатпаева, Сарканд, Саркандский район, Алма...  
12992  (21, проспект Республики, район "Сарыарка", Ну...  
13531                                               None  
11837  (ЖК "Кайнар", 28, улица Куйши Дина, район "Алм...  
13615                                               None

In [292]:
bin_sample_sample = bin_sample_sample.dropna()

In [293]:
bin_sample_sample['lat'] = bin_sample_sample['location'].apply(lambda x: x.latitude)
bin_sample_sample['lon'] = bin_sample_sample['location'].apply(lambda x: x.longitude)

<ipython-input-293-62cda047b232>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_sample['lat'] = bin_sample_sample['location'].apply(lambda x: x.latitude)
<ipython-input-293-62cda047b232>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_sample['lon'] = bin_sample_sample['location'].apply(lambda x: x.longitude)


In [297]:
bin_sample_sample['location'][1]

Location(улица Нурсултана Назарбаева, Петропавловск Г.А., Северо-Казахстанская область, 150000, Қазақстан, (54.8732011, 69.1380132, 0.0))

In [298]:
bin_sample_sample.sample(20)

bin                                           full_add  \
1880  020640000976                 Алматы,район Медеуский, Пушкина,60   
386   000615550151  обл. Восточно-Казахстанская, Усть-каменогорск,...   
1793  020440006581                                   ТАРАЗ,Толе би,22   
1010  010440000530                                  Семей, Герцена,13   
555   000940000894                                  КОСТАНАЙ,Речная,3   
1972  020740004863               Алматы,район Турксибский,Бехтерева,1   
770   001240005694                           КАРАГАНДА,Молокова,119/1   
452   000716650731           Алматы,район Алмалинский,Макатаева,127/1   
189   000340000051                                   ТАРАЗ,Толе Би,18   
1395  011040003632             Алматы,район Алатауский, Рыскулова,103   
353   000540002221            Шымкент,район Енбекшинский,Алдиярова,60   
1638  020240000644                                Алматы,Чехова,10/11   
1422  011041006375                                Павлодар,Астана,119   
1127  010626550202  обл. Павлодарская, Экибастуз, МУХТАРА АУЭЗОВА,...   
1256  010840000076                     УСТЬ-КАМЕНОГОРСК,  Абая, 100/2   
1840  020540003631                                АСТАНА,Жиенкулова,5   
1515  011240004360                            АСТАНА,Абылай хана,41/3   
191   000340000357                      район Алмалинский,Толе Би,214   
1835  020540002790                               МЕДЕУСКИЙ,Пушкина,13   
588   001002650128               обл. Атырауская, Атырау,Гурьевская,4   

                                               location        lat        lon  
1880  (60, улица Пушкина, Медеуский район, Алматы, 0...  43.253106  76.954255  
386   (13, Ауэзова проспект, Речной Вокзал, Стрелка,...  49.948380  82.612601  
1793  (Толе би, Тараз, Тараз Г.А., Жамбылская област...  42.900782  71.380777  
1010  (13, улица Герцена, 72-й квартал, Семей, Город...  50.412176  80.242413  
555   (3, Речная улица, Горняцкая поселковая админис...  52.999021  63.316528  
1972  (1, улица Бехтерева, Турксибский район, Алматы...  43.322708  76.942211  
770   (улица Молокова, Караганда, Караганда Г.А., Ка...  49.774522  73.086461  
452   (Jewelry Magic, 127/1, улица Макатаева, Алмали...  43.263670  76.929064  
189   (Толе Би, Тараз, Тараз Г.А., Жамбылская област...  42.901393  71.334249  
1395  (Алатау, ТЦ, 103/3, проспект Рыскулова, Самғау...  43.266938  76.869197  
353   (60, улица Алдиярова, Енбекшинский район, Шымк...  42.306299  69.626209  
1638  (Чехова, Алма Тау, Медеуский район, Бесагаш, Т...  43.246639  77.085619  
1422  (119, улица Астана, Химгородки, Павлодар, Павл...  52.284086  76.939554  
1127  (улица Мухтара Ауэзова, Экибастуз, Экибастуз Г...  51.726414  75.308956  
1256  (Абая проспект, Стрелка, Усть-Каменогорск, Вос...  49.966701  82.616648  
1840  (5, улица Жиенкулова, район "Байконур", Нур-Су...  51.174874  71.458853  
1515  (41/3, проспект Абылай Хана, район "Алматы", Н...  51.154231  71.492404  
191   (214, улица Толе Би, Алмалинский район, Алматы...  43.250394  76.888981  
1835  (13, улица Пушкина, Медеуский район, Алматы, A...  43.267845  76.950713  
588   (4, Гурьевская улица, Атырау, Атырау Г.А., Аты...  47.102736  51.925692

In [179]:
bin_sample_notna = bin_sample[bin_sample['location'].notna()]

In [180]:
bin_sample_notna.shape

(258, 3)

In [187]:
bin_sample_isna = bin_sample[bin_sample['location'].isna()]

In [200]:
bin_sample_isna.shape

(242, 3)

<ipython-input-189-fc9addbd22bb>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample_isna["full_add"] = bin_sample_isna["full_add"].str.replace(r'\пр+\.', "")
<ipython-input-189-fc9addbd22bb>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna["full_add"] = bin_sample_isna["full_add"].str.replace(r'\пр+\.', "")


In [190]:
bin_sample_isna['location'] = bin_sample_isna['full_add'].apply(lambda x: geolocator.geocode(x))

<ipython-input-190-194f14088456>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna['location'] = bin_sample_isna['full_add'].apply(lambda x: geolocator.geocode(x))


In [201]:
bin_sample_isna_2 = bin_sample_isna[bin_sample_isna['location'].isna()]

In [202]:
bin_sample_isna_2.shape

(215, 3)

<ipython-input-207-bea1ff87b02e>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample_isna_2["full_add"] = bin_sample_isna_2["full_add"].str.replace(r'\д+\.', "")
<ipython-input-207-bea1ff87b02e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_2["full_add"] = bin_sample_isna_2["full_add"].str.replace(r'\д+\.', "")


In [208]:
bin_sample_isna_2['location'] = bin_sample_isna_2['full_add'].apply(lambda x: geolocator.geocode(x))

<ipython-input-208-debddca887ff>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_2['location'] = bin_sample_isna_2['full_add'].apply(lambda x: geolocator.geocode(x))


In [220]:
bin_sample_isna_3 = bin_sample_isna_2[bin_sample_isna_2['location'].isna()]

In [223]:
bin_sample_isna_3.sample(20)

bin                                           full_add location
300  000440005788                 Астана,ул Ауэзова,д 40 кв офис 208     None
177  000317501054  обл. Алматинская,с. Жанатурмыс,Абылхайыр хана,41А     None
460  000740000546                    СЕМЕЙ, Кайыма Мухамедханова, 49     None
328  000522550186  обл. Карагандинская,район Имени казыбек би,Ала...     None
101  000223550041               обл. Костанайская, Житикара,6 мкр,62     None
277  000440001933  обл. Западно-Казахстанская,с. Шынгырлау, Бокен...     None
319  000501501689                   район Бостандыкский,Джандосова,1     None
201  000340001553            Алматы,район Турксибский, Д. Бедного, 5     None
346  000540001500  обл. Акмолинская, Кокшетау, Кенесары Касымова,...     None
362  000540004139                             САРЫАҒАШ,Ташкентская,7     None
114  000240001549   УСТЬ-КАМЕНОГОРСК,имени Александра Протозанова,17     None
447  000711600956                    ШЫМКЕНТ,Жангельдина дом 16,кв16     None
238  000340007847  обл. Западно-Казахстанская,с. Мичуринское,Альк...     None
49   000140002415         Павлодар,ПРОМЫШЛЕННАЯ ЗОНА ЦЕНТРАЛЬНАЯ,406     None
282  000440002505                           с. Әйет,им. Б.Майлина,23     None
391  000620550505      Нур-Султан,район Есиль,улица Роза Бағланова,4     None
200  000340001503                                 АСТАНА, Гете д,3/2     None
490  000808600021                  Алматы,район Ауэзовский,мкр.4,10Б     None
372  000540005741                                с. Кулан,Болысова,1     None
227  000340006643             АКТАУ,29мкр.,,строительный рынок"Адал"     None

<ipython-input-226-55316a88407b>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample_isna_3["full_add"] = bin_sample_isna_3["full_add"].str.replace('мкр.', "микрорайон ", case=False)
<ipython-input-226-55316a88407b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_3["full_add"] = bin_sample_isna_3["full_add"].str.replace('мкр.', "микрорайон ", case=False)


In [227]:
bin_sample_isna_3['location'] = bin_sample_isna_3['full_add'].apply(lambda x: geolocator.geocode(x))

<ipython-input-227-fcef5d67e1d4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_3['location'] = bin_sample_isna_3['full_add'].apply(lambda x: geolocator.geocode(x))


In [233]:
bin_sample_isna_4 = bin_sample_isna_3[bin_sample_isna_3['location'].isna()]

In [234]:
bin_sample_isna_4

bin                                           full_add location
0    000103501685                           КЫЗЫЛ ЖУЛДЫЗ,Апатаева,16     None
2    000105651311  обл. Карагандинская,район имени Казыбек Би,Бух...     None
10   000112550593  обл. Восточно-Казахстанская,с. Урджар,ЖАНДОСОВ,37     None
16   000117551452           обл. Акмолинская,пос. Аршалы,Писарева,14     None
17   000117650715                            АКТОБЕ,Бр.Жубановых,285     None
..            ...                                                ...      ...
489  000808500987   Алматы,район Наурызбайский,микрорайон  Акжар ...     None
490  000808600021           Алматы,район Ауэзовский,микрорайон 4,10Б     None
493  000812000174  обл. Алматинская,с. Балпык би,Кабанбай батыра,...     None
495  000817601009            Нур-Султан,район Сарыарка,Республика,19     None
498  000827551330                 обл. Акмолинская,с. Турген,Есил,12     None

[194 rows x 3 columns]

<ipython-input-238-37fdb8680dd5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_4['full_add'] = bin_sample_isna_4['full_add'].apply(lambda x: re.sub(r'(\d)\D+$', r'\1', x))


In [239]:
bin_sample_isna_4['location'] = bin_sample_isna_4['full_add'].apply(lambda x: geolocator.geocode(x))

<ipython-input-239-de19b5ad38ee>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna_4['location'] = bin_sample_isna_4['full_add'].apply(lambda x: geolocator.geocode(x))


In [244]:
bin_sample_isna_4[bin_sample_isna_4['location'].isna()]

bin                                           full_add location
0    000103501685                           КЫЗЫЛ ЖУЛДЫЗ,Апатаева,16     None
2    000105651311  обл. Карагандинская,район имени Казыбек Би,Бух...     None
10   000112550593  обл. Восточно-Казахстанская,с. Урджар,ЖАНДОСОВ,37     None
16   000117551452           обл. Акмолинская,пос. Аршалы,Писарева,14     None
17   000117650715                            АКТОБЕ,Бр.Жубановых,285     None
..            ...                                                ...      ...
486  000805601559               Шымкент,район Каратау, Тулеметова,69     None
489  000808500987   Алматы,район Наурызбайский,микрорайон  Акжар ...     None
493  000812000174  обл. Алматинская,с. Балпык би,Кабанбай батыра,...     None
495  000817601009            Нур-Султан,район Сарыарка,Республика,19     None
498  000827551330                 обл. Акмолинская,с. Турген,Есил,12     None

[182 rows x 3 columns]

In [131]:
bin_sample_notna['lat'] = bin_sample_notna['location'].apply(lambda x: x.latitude)
bin_sample_notna['lon'] = bin_sample_notna['location'].apply(lambda x: x.longitude)

<ipython-input-131-add25c768bb9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_notna['lat'] = bin_sample_notna['location'].apply(lambda x: x.latitude)
<ipython-input-131-add25c768bb9>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_notna['lon'] = bin_sample_notna['location'].apply(lambda x: x.longitude)


In [132]:
bin_sample_isna = bin_sample[bin_sample['location'].isna()]

In [140]:
bin_sample_isna.sample(20)

bin                                           full_add location
83   000140006993                           с. Кен,Тбаева курылыс,2А     None
103  000224600324                Алматинская, Талдыкорн,Акын сара,84     None
193  000340000664                           Семей,перек заводской,10     None
123  000240002964          Акмолинская,с. Новокубанка,Кан Де Хана,2А     None
110  000240000715  АЛМАТЫ,Илийский район,с.Отен батыра,Куат Тауел...     None
32   000140000240                  Нур-Сан,район Байконыр,Иманова,12     None
397  000640000030                          САУМАЛКОЛЬ,АЗС Саумалколь     None
61   000140004907                                 КОКШЕТАУ,Гарина,53     None
169  000311651173                  Костанайская, Костанай,пр.Абая,94     None
382  000605551343                     Нур-Сан,район Есиль,Сынак,21/1     None
427  000640003616                             АЛМАТЫ,Станиславско,77     None
279  000440002426                                АКТОБЕ,Смалова,д.9Б     None
109  000240000501                               Алматы,М-он 2,д.13 а     None
214  000340004984              Алматы,район Ауэзовский,мкр. 8,д. 4 А     None
274  000440001814              Костанайская,с. Октябрьское,Лесная,18     None
67   000140005678                                  ЖЕТЫСАЙ,Яссауй,69     None
381  000601650637   Восточно-Казахстанская, Усть-Каменорск,Подсне...     None
5    000108550224                      Акмолинская,с. Акмол,Гарина,6     None
266  000440000876                              Оскемен, Поаничная,51     None
27   000127601781                   Нур-Сан,район Есиль,Бокейхана,15     None

In [138]:
bin_sample_isna["full_add"] = bin_sample_isna["full_add"].str.replace('ул.', "",case=False)

<ipython-input-138-59c7cf9f5de5>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  bin_sample_isna["full_add"] = bin_sample_isna["full_add"].str.replace('ул.', "",case=False)
<ipython-input-138-59c7cf9f5de5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna["full_add"] = bin_sample_isna["full_add"].str.replace('ул.', "",case=False)


In [141]:
bin_sample_isna['location'] = bin_sample_isna['full_add'].apply(lambda x: geolocator.geocode(x))

<ipython-input-141-194f14088456>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bin_sample_isna['location'] = bin_sample_isna['full_add'].apply(lambda x: geolocator.geocode(x))


In [142]:
bin_sample_isna

bin                                           full_add location
0    000103501685                              КЫЗЫЛ ЖЫЗ,Апатаева,16     None
2    000105651311   Карандинская,район имени Казыбек Би,Бухар-Жыр...     None
3    000106600917       Алматы,район Алмалинский,Бонбай Батыра,д.300     None
4    000107500103         Алматы,район Медеуский,пр.Назарбаева,д.248     None
5    000108550224                      Акмолинская,с. Акмол,Гарина,6     None
..            ...                                                ...      ...
494  000814601607             Алматы,район Ауэзовский,мкр. 8,д. 37/1     None
495  000817601009               Нур-Сан,район Сарыарка,Республика,19     None
496  000821650039   Восточно-Казахстанская, Усть-Каменорск, Казах...     None
498  000827551330                        Акмолинская,с. Турн,Есил,12     None
499  000840000137                Алматинская, Талдыкорн,Жансурова,78     None

[295 rows x 3 columns]

- из 500 улиц распознаванию поддались 170 улиц, среди которых также могут быть ошибки (baseline)
- после изменения р-н на район и удаления г. количество найденных точек возросло до 258
- после удаления улица и просппект возросло до 285
- удаление д. привело к увеличению поиска до 299
- замена мкр на микрорайон увеличила до 306
- после удаления букв после номера дома цифра достигла 318